8160821: VarHandle accesses are penalized when argument conversion is required

liach opened the pull request:
Since access descriptor is created for each VH operation site, we can optimistically cache the adapted method handle in a site if the site operates on a constant VH.  Used a C2 IR test to verify such a setup through an inexact VarHandle invocation can be constant folded through (previously, it was blocked by `asType`)

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- \[x\] Change must not contain extraneous whitespace
- \[x\] Commit message must refer to an issue
- \[ \] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Error
&nbsp;⚠️ Pull request body is missing required line: `- [x] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai).`

### Issue
 * [JDK-8160821](https://bugs.openjdk.org/browse/JDK-8160821): VarHandle accesses are penalized when argument conversion is required (**Enhancement** - P3)


### Reviewers
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - **Reviewer**) 🔄 Re-review required (review applies to [d734e8a6](https://git.openjdk.org/jdk/pull/28585/files/d734e8a61b3917e57bee12ca85582b64ff681123))
 * [John R Rose](https://openjdk.org/census#jrose) (@rose00 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/28585/head:pull/28585` \
`$ git checkout pull/28585`

Update a local copy of the PR: \
`$ git checkout pull/28585` \
`$ git pull https://git.openjdk.org/jdk.git pull/28585/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 28585`

View PR using the GUI difftool: \
`$ git pr show -t 28585`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/28585.diff">https://git.openjdk.org/jdk/pull/28585.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/28585#issuecomment-3598695468)
</details>

liach committed:
8160821

liach committed:
Logical fallacy

liach committed:
Tweak VH usage in some classes

liach commented on src/java.base/share/classes/java/lang/invoke/VarHandle.java:
```diff
@@ -2016,15 +2016,35 @@ static final class AccessDescriptor {
         }
+
+        @ForceInline
+        MethodHandle adaptedMethodHandle(VarHandle vh) {
```
Indeed, I should move the adaptedMh read into `constant == MethodHandleImpl.CONSTANT_YES` block.

`constant != MethodHandleImpl.CONSTANT_NO` prevents capturing any further if the VH is known non-constant; we keep this branch in constant case in case the adapted MH is not ready when we know the VH is constant.

liach replied:
So an `AccessDescriptor` is created for each sigpoly VH site in the source code. Usually it is `VH.operation()`, but it is legal to use a non-constant VarHandle variable and call an operation on that. If `constant == MethodHandleImpl.CONSTANT_NO`, we are sure that we have the non-constant case, so we cannot trust that cached method handle, and there is no point further caching. We can only read that previous MH conversion cache if `constant == MethodHandleImpl.CONSTANT_YES` because this means our cache is always correct.

iwanowww replied:
I still find it confusing, especially tri-state logic part.

For background, `isCompileConstant` was introduced as part of LF sharing effort to get rid of Java-level profiling in optimized code. The pattern is was designed for was:
```
  if (isCompileConstant(...)) {
    return ...;
  } else {
    ... // do some extra work (either in interpreter, C1, or not-fully-optimized version in C2) 
  }
```

In this patch, you don't follow that pattern and also add new state (`CONSTANT_PENDING`) to distinguish interpreter/C1 from C2. What's the motivation? Why do you want to avoid cache updates coming from C2-generated code?

iwanowww replied:
Sorry, I still don't understand how it is intended to work. Why does `MethodHandleImpl.isCompileConstant(vh) == true` imply that the cached value is compatible with the constant `vh`?

```
// Keep capturing - vh may suddenly get promoted to a constant by C2
```

Capturing happens outside compiler thread. It is not affected by C2 (except when it completely prunes the whole block).

So, either any captured adaptation is valid/compatible or there's a concurrency issue when C2 kicks in and there's a concurrent cache update happening with incompatible version.

liach replied:
> any captured adaptation is valid/compatible

Yes, if `vh` is a constant, any captured adaptation from `vh.getMethodHandle(mode).asType(symbolicMethodTypeInvoker)` is valid/compatible.

For thread safety, MethodHandle supports safe publication, so I think we are fine publishing this way.

JornVernee replied:
Looking at this, I'm not sure we can assume that we only see one mode and type when the VH is constant. There seems to be a lot of non-local reasoning involved.

For example, you could have a var handle invoker created with `MethodHandless::varHandleInvoker`, which is cached, so the `AccessDescriptor` can be shared among many different use sites. For an individual use-site, the receiver VH may well be a constant, but that doesn't mean that the cache isn't polluted by the var handle from another use site, as far as I can tell.

The thread safety issue comes from a C2 thread racing to read the `lastAdaption` cache vs another Java thread writing to the cache. AFAICS, this race is still possible even when `vh` is a compile time constant.

liach committed:
Review tweaks

liach commented on src/java.base/share/classes/java/lang/invoke/VarHandle.java:
```diff
@@ -2016,15 +2016,35 @@ static final class AccessDescriptor {
+        MethodHandle adaptedMethodHandle(VarHandle vh) {
+            var constant = MethodHandleImpl.isCompileConstant(vh);
+            var cache = adaptedMh;
+            if (constant == MethodHandleImpl.CONSTANT_YES && cache != null) {
```
MethodHandle is immutable and can be safely published. So this is ok.

liach replied:
In extreme cases where a barrier is needed, java.lang.invoke already issue necessary barriers, most notably the storeStoreFence, such as https://github.com/openjdk/jdk/blob/135661b4389663b8c2e348d9e61e72cc628636bb/src/java.base/share/classes/java/lang/invoke/CallSite.java#L138
or https://github.com/openjdk/jdk/blob/135661b4389663b8c2e348d9e61e72cc628636bb/src/java.base/share/classes/java/lang/ClassValue.java#L411-L417

iwanowww commented on src/java.base/share/classes/java/lang/invoke/VarHandle.java:
```diff
@@ -2016,15 +2016,35 @@ static final class AccessDescriptor {
+
+            // This is still a hot path if vh is not constant - in this case,
+            // asType is the bottleneck for constant folding, unfortunately
+            var result = vh.getMethodHandle(mode).asType(symbolicMethodTypeInvoker);
```
`mode` and `symbolicMethodTypeInvoker` are part of `AccessDescriptor` while `vh` comes as an argument. What guarantees that a cached adapter is compatible with `vh` observed during subsequent calls? It means that `vh` shape stays exactly the same shape. Is it correct? Would be good to have it validated with asserts.

liach replied:
I am assuming that the previous `vh` observed is compatible with future ones if compiler can fold the `vh` into a constant. If it is not, we can drop the updates to the cache field in the C2 compiled slow code.

liach committed:
Stage

liach committed:
Merge branch 'master' of https://github.com/openjdk/jdk into fix/vh-adapt-cache

liach committed:
Redundant change

liach committed:
Rollback getAndAdd for now

liach commented:
After consulting with @iwanowww, I realized the non-constant status cannot be determined, that the C2 compiled method can even transition from 0 to 1, so I am simplifying this code to only handle the constant case. It seems the getAndAdd IR test no longer fails with this change, and I removed a lot of other redundant changes.

I updated the VarHandleExact benchmark added by @JornVernee, and added a case of dropping return values by changing access mode to `getAndAdd` consistently. Now they have the following performance numbers:
```
Benchmark                                        Mode  Cnt  Score   Error  Units
VarHandleExact.exact_exactInvocation             avgt   30  3.843 ± 0.062  ns/op
VarHandleExact.generic_exactInvocation           avgt   30  3.797 ± 0.049  ns/op
VarHandleExact.generic_genericInvocation         avgt   30  3.757 ± 0.034  ns/op
VarHandleExact.generic_returnDroppingInvocation  avgt   30  3.754 ± 0.026  ns/op
```

iwanowww commented on test/micro/org/openjdk/bench/java/lang/invoke/VarHandleExact.java:
```diff
@@ -69,16 +69,21 @@ public void setup() {
+    }
+
+    @Benchmark
+    public void generic_returnDroppingInvocation() {
```
What about "all-generic" case (` { generic.getAndAdd(data, 42); }`)?

liach committed:
Fix problem identified by Jorn

liach committed:
Copyright years

liach committed:
Test from Jorn

liach committed:
Revert void special case removal due to C2 shortage causing TestZGCBarrierElision::testAtomicThenAtomicAnotherField failure

iwanowww commented on src/java.base/share/classes/java/lang/invoke/VarHandle.java:
```diff
@@ -2017,14 +2017,49 @@ static final class AccessDescriptor {
+        // Object indirection is the only way to ensure the vh and mh are not
+        // from two writes (they must not be tearable)
+        private record Adaption(VarHandle vh, MethodHandle mh) {}
+        private @Stable Adaption adaption;
```
Then it makes sense to limit the caching to safe cases only for now. Otherwise, it would functionally regress due to a possible memory leak.

JornVernee reviewed (commented):
Latest version looks much better to me (as mentioned offline). What was the issue with the failing test around the removal of the _V guard template?

Also, looks like the new IR test is failing in GHA

liach committed:
Merge branch 'master' of https://github.com/openjdk/jdk into fix/vh-adapt-cache

liach committed:
Try to avoid loader leak

liach committed:
Bugs and verify loader leak

liach committed:
Merge branch 'master' of https://github.com/openjdk/jdk into fix/vh-adapt-cache

liach committed:
Review

liach commented on src/java.base/share/classes/java/lang/invoke/IndirectVarHandle.java:
```diff
@@ -105,4 +105,12 @@ MethodHandle getMethodHandleUncached(int mode) {
+        // Ideally we should check the MH produced by handleFactory too,
+        // but checking the signature type of MH mostly works
+        return MethodHandle.isReachableFrom(vform.getMethodType(0), cl)
+                && target.isReachableFrom(cl);
```
It is, but after all, there can be similar resource leak risks from calling LambdaMetafactory or other APIs too. We expect AccessDescriptor construction to be mostly limited to specific sites.

eme64 commented on test/hotspot/jtreg/compiler/c2/irTests/constantFold/VarHandleMismatchedTypeFold.java:
```diff
@@ -0,0 +1,77 @@
+ * @bug 8160821
+ * @summary Verify constant folding is possible for mismatched VarHandle access
+ * @library /test/lib /
+ * @requires vm.compiler2.enabled
```
What would happen if you removed this? Is the test expected to fail anywhere? IR rules are executed only if C2 is available anyway, so probably you don't need this restriction.

eme64 reviewed (changes requested):
Was asked on slack to review IR tests, so that's all I'm doing here - won't review the whole patch ;)

iwanowww commented on make/jdk/src/classes/build/tools/methodhandle/VarHandleGuardMethodGenerator.java:
```diff
@@ -122,11 +122,14 @@ final class VarHandleGuards {
+    // Currently this void version is needed because otherwise
+    // TestZGCBarrierElision.testAtomicThenAtomicAnotherField fails
+    // However, testArrayAtomicThenAtomic, testAtomicThenAtomic, and
+    // testArrayAtomicThenAtomicAtUnknownIndices works
```
It doesn't look right. Was is the root cause of the failure? Can it be a test bug?

liach committed:
Merge branch 'master' of https://github.com/openjdk/jdk into fix/vh-adapt-cache

liach committed:
Stage

liach committed:
Merge branch 'master' of https://github.com/openjdk/jdk into fix/vh-adapt-cache

liach committed:
Missed IR test review, rearrange benches

liach commented:
Since I just noticed the actual cause of the failure of caching is that AD is created per name+type combination, I have created a benchmark case for that instead of reusing the existing exact ones:

```
Benchmark                                 Mode  Cnt  Score   Error  Units
VarHandleTypeMismatch.exactInvocation     avgt   30  0.396 ± 0.009  ns/op
VarHandleTypeMismatch.genericInvocation   avgt   30  0.375 ± 0.009  ns/op
VarHandleTypeMismatch.pollutedInvocation  avgt   30  8.281 ± 0.222  ns/op
```

liach committed:
Merge branch 'master' of https://github.com/openjdk/jdk into fix/vh-adapt-cache

liach committed:
Design detail updates, thanks to jorn

liach committed:
Wording update, thanks Jorn

rose00 reviewed (approved):
Good work.  I’m assuming you will address my previous comments.

liach commented:
/open

liach commented:
/open
